File: Data Structures & Algorithms/decode-ways/submission-2.py
class Solution:
    def numDecodings(self, s: str) -> int:
        dp={}
        def f(i):
            if i in dp:return dp[i]
            if i==len(s):return 1
            if s[i]=='0':return 0
            res=f(i+1)
            if i+1<len(s) and (s[i] =='1' or s[i]=='2' and s[i+1] in '0123456'):
                res+=f(i+2)
            dp[i]=res
            return res
        return f(0)

File: Data Structures & Algorithms/string-encode-and-decode/submission-2.js
class Solution {
  /**
   * @param {string[]} strs
   * @returns {string}
   */
  encode(strs) {
    let res = "";
    for (const s of strs) {
      res += s.length + "#" + s;
    }
    return res;
  }

  /**
   * @param {string} str
   * @returns {string[]}
   */
  decode(str) {
    const arr = str.split("");
    const res = [];
    while (arr.length) {
      const sublen = arr.splice(0, arr.indexOf("#")).join("");
      arr.shift();
      const substr = arr.splice(0, sublen).join("");
      res.push(substr);
    }
    return res;
  }
}

